src/pyroboplan/ik/nullspace_components.py

```diff
@@ -74,3 +75,100 @@ def joint_center_nullspace_component(model, q, gain=1.0):
     """
     joint_center_positions = 0.5 * (model.lowerPositionLimit + model.upperPositionLimit)
     return gain * (joint_center_positions - q)
+
+
+def collision_avoidance_nullspace_component(
+    model,
+    data,
+    collision_model,
+    collision_data,
+    q,
+    dist_padding=0.05,
+    max_vel=1.0,
+    damping=0.0001,
+    gain=1.0,
+):
+    """
+    Returns a collision avoidance nullspace component.
+
+    Parameters
+    ----------
+        model : `pinocchio.Model`
+            The model from which to generate a random state.
+        data : `pinocchio.Data`
+            The model data to use for collision distance checks.
+        collision_model : `pinocchio.GeometryModel`
+            The model with which to check collision distances.
+        collision_data : `pinocchio.GeometryData`
+            The collision model data to use for collision distance checks.
+        q : array-like
+            The joint configuration for the model.
+        dist_padding : float
+            The distance padding, in meters, on the collision distances.
+            For example, a distance padding of 0.1 means collisions have an influence 10 cm away from actual collision.
+        max_vel : float
+            The maximum velocity norm that can be returned by this component.
+        damping : float
+            Damping value, between 0 and 1, for the collision Jacobian pseudoinverse.
+        gain : float, optional
+            A gain to modify the relative weight of this term.
+
+    Returns
+    -------
+        array-like
+            An array containing the collision avoidance nullspace terms.
+    """
+    coll_component = np.zeros_like(model.lowerPositionLimit)
+
+    # Find all the collision distances at the current state.
+    pinocchio.computeDistances(model, data, collision_model, collision_data, q)
+
+    # For each collision pair within a distance threshold, calculate its collision Jacobian
+    # and use it to push the corresponding joint values away from collision.
+    for cp, dr in zip(collision_model.collisionPairs, collision_data.distanceResults):
+        if dr.min_distance > dist_padding:
```

Comment:
Should this also check if the distance is 0 (ie, in collision)? Otherwise I think things break down?

Reply:
I need to check again, but I thought the distances were negative if the objects are overlapping

Reply:
I checked again. They are not. So now I am actually doing a collision check as well and getting the normal/penetration depth from there.

It's unclear to me if I did it right, but it's a step in the right direction...